pkg/skaffold/test/custom/custom.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package custom
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/list"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+type Runner struct {
+	customTest     latest.CustomTest
+	testWorkingDir string
+	localDaemon    docker.LocalDaemon
+}
+
+// New creates a new custom.Runner.
+func New(cfg docker.Config, wd string, ct latest.CustomTest) (*Runner, error) {
+	localDaemon, err := docker.NewAPIClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Runner{
+		customTest:     ct,
+		testWorkingDir: wd,
+		localDaemon:    localDaemon,
+	}, nil
+}
+
+// Test is the entrypoint for running custom tests
+func (ct *Runner) Test(ctx context.Context, out io.Writer, _ []build.Artifact) error {
+	if err := ct.runCustomCommand(ctx, out); err != nil {
+		return fmt.Errorf("running custom test command: %w", err)
+	}
+
+	return nil
+}
+
+func (ct *Runner) runCustomCommand(ctx context.Context, out io.Writer) error {
+	test := ct.customTest
+
+	// Expand command
+	command, err := util.ExpandEnvTemplate(test.Command, nil)
+	if err != nil {
+		return fmt.Errorf("unable to parse test command %q: %w", test.Command, err)
+	}
+
+	logrus.Infof("Running custom test command %q", command)
+	if test.TimeoutSeconds <= 0 {
+		color.Default.Fprintf(out, "Running custom test command: %q\n", command)
+	} else {
+		color.Default.Fprintf(out, "Running custom test command: %q with timeout %d s\n", command, test.TimeoutSeconds)
+		newCtx, cancel := context.WithTimeout(ctx, time.Duration(test.TimeoutSeconds)*time.Second)
+		defer cancel()
+		ctx = newCtx
+	}
+
+	var cmd *exec.Cmd
+	// We evaluate the command with a shell so that it can contain env variables.
+	if runtime.GOOS == "windows" {
+		cmd = exec.CommandContext(ctx, "cmd.exe", "/C", command)
+	} else {
+		cmd = exec.CommandContext(ctx, "sh", "-c", command)
+	}
+
+	cmd.Stdout = out
+	cmd.Stderr = out
+	cmd.Env = ct.env()
+
+	if err := util.RunCmd(cmd); err != nil {
```

Comment:
Should we use `util.RunCmdOut` to capture the test logs and make it user friendly?
1) stream test script logs as they become avaliable

Reply:
I am also thinking if we shd switch to using `exec.Start` and graceful termination on cancelling like 
https://github.com/GoogleContainerTools/skaffold/blob/dafb4152be74c7cd7d363790f2099f2b6f057ac3/pkg/skaffold/build/custom/script.go#L41

Reply:
I considered graceful termination but current implementation of graceful termination doesn't accommodate timer.

Reply:
hmm so right now, users will only know if a command failed or not? The DD mentions, skaffold will be logging output from the command https://docs.google.com/document/d/1aMfQtlf1gEFbMMVbWy2Iep9WQalLHou6Myc9EVbz01w/edit?ts=6007c302&resourcekey=0-T7xncfS8izrFHTaMeFtMfg#heading=h.7008ql63pa1

We need to support this.

Reply:
I was originally using exec.Start() but updated to use exec.Run() per this previous feedback
https://github.com/GoogleContainerTools/skaffold/pull/5451#discussion_r584945883

Reply:
using `util.RunCmd` is not going to show the command output right?
can you verify this and update the PR description?

Reply:
`util.RunCmd` does show the command output.